File: ios/Classes/LeancloudQuery.m
//
//  LeancloudQuery.m
//  leancloud_flutter_plugin
//
//  Created by Luna on 2019/1/7.
//

#import "LeancloudQuery.h"
#import <AVOSCloud/AVOSCloud.h>
#import "LeancloudArgsConverter.h"

@implementation LeancloudQuery

const NSString *QUERYMETHOD = @"queryMethod";
const NSString *ARG1 = @"arg1";
const NSString *ARG2 = @"arg2";

- (void) query:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSDictionary *avQueryJson = [LeancloudArgsConverter getAVQueryJsonObject:call result:result];
    NSString *className = avQueryJson[@"className"];
    NSString *queriesString = avQueryJson[@"queries"];
    NSDictionary *queriesJson = [LeancloudArgsConverter stringToJson:queriesString];
    AVQuery *query = [AVQuery queryWithClassName:className];
    for (NSDictionary *oneQuery in queriesJson) {
        if([oneQuery[@"queryMethod"] isEqualToString:@"get"]) { // if has getObjectWithId function
            [query getObjectInBackgroundWithId:oneQuery[@"arg1"] block:^(AVObject *object, NSError *error) {
                NSMutableDictionary *serializedJSONDictionary = [object dictionaryForObject];
                NSError *err;
                NSData *jsonData = [NSJSONSerialization dataWithJSONObject:serializedJSONDictionary options:0 error:&err];
                NSString *serializedString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
                result(serializedString);
                return;
            }];
        }
        if([oneQuery[QUERYMETHOD] isEqualToString:@"equalTo"]) {
            [query whereKey:oneQuery[ARG1] equalTo:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"notEqualTo"]) {
            [query whereKey:oneQuery[ARG1] notEqualTo:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"greaterThan"]) {
            [query whereKey:oneQuery[ARG1] greaterThan:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"greaterThanOrEqualTo"]) {
            [query whereKey:oneQuery[ARG1] greaterThanOrEqualTo:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"lessThan"]) {
            [query whereKey:oneQuery[ARG1] lessThan:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"lessThanOrEqualTo"]) {
            [query whereKey:oneQuery[ARG1] lessThanOrEqualTo:oneQuery[ARG2]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"orderByAscending"]) {
            [query orderByAscending:oneQuery[ARG1]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"orderByDescending"]) {
            [query orderByDescending:oneQuery[ARG1]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"addAscendingOrder"]) {
            [query addDescendingOrder:oneQuery[ARG1]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"addDescendingOrder"]) {
            [query addDescendingOrder:oneQuery[ARG1]];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"limit"]) {
            query.limit = [oneQuery[ARG1] intValue];
        } else if([oneQuery[QUERYMETHOD] isEqualToString:@"skip"]) {
            query.skip = [oneQuery[ARG1] intValue];
        } else {
            result([FlutterError errorWithCode:@"params-error"
                                       message:@"no such query queryMethod name, please check again"
                                       details:[NSString stringWithFormat:@"no such query queryMethod name, %@", oneQuery[QUERYMETHOD]]]);
        }
    }
    [query findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
        if (error != nil) {
            result([FlutterError errorWithCode:@"leancloud-error"
                                       message:[error localizedDescription]
                                       details:[error localizedFailureReason]]);
        } else {
            NSMutableArray *array = [[NSMutableArray alloc] init];
            for (AVObject *object in objects) {
                NSMutableDictionary *serializedJSONDictionary = [object dictionaryForObject];
                NSError *err;
                NSData *jsonData = [NSJSONSerialization dataWithJSONObject:serializedJSONDictionary options:0 error:&err];
                NSString *serializedString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
                [array addObject:serializedString];
            }
            NSDictionary *jsonDictionary = [NSDictionary dictionaryWithObjectsAndKeys:
                                     array, @"objects",
                                     nil];
            NSData* jsonData =[NSJSONSerialization dataWithJSONObject:jsonDictionary options:NSJSONWritingPrettyPrinted error:&error];
            NSString* jsonString =[[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
            result(jsonString);
        }
    }];
}

@end

File: android/src/main/java/com/lunagao/leancloudflutterplugin/LeancloudQuery.java
package com.lunagao.leancloudflutterplugin;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

import java.util.List;

import cn.leancloud.AVObject;
import cn.leancloud.AVQuery;
import cn.leancloud.query.AVCloudQueryResult;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

class LeancloudQuery {

    /**
     * query AVObject
     * More detail please check: https://github.com/leancloud/java-sdk-all/wiki/1.%E5%AD%98%E5%82%A8-3-AVQuery
     *
     * function:
     *  get
     *  whereEqualTo
     *  whereNotEqualTo
     *  whereGreaterThan
     *  whereGreaterThanOrEqualTo
     *  whereLessThan
     *  whereLessThanOrEqualTo
     *  whereStartsWith
     *  whereContains
     *  whereMatches
     *  limit
     *  skip
     *
     * @param call MethodCall from LeancloudFlutterPlugin.onMethodCall function
     * @param result MethodChannel.Result from LeancloudFlutterPlugin.onMethodCall function
     */
    void query(MethodCall call, final MethodChannel.Result result) {
        JSONObject avQueryJson = LeancloudArgsConverter.getAVQueryJsonObject(call, result);
        assert avQueryJson != null;
        String className = avQueryJson.getString("className");
        String queriesString = avQueryJson.getString("queries");
        JSONArray queriesJson = JSON.parseArray(queriesString);
        AVQuery<AVObject> avQuery = new AVQuery<>(className);
        for (Object oneQueryObject : queriesJson) {
            JSONObject oneQuery;
            if (oneQueryObject instanceof JSONObject) {
                oneQuery = (JSONObject) oneQueryObject;
            } else {
                oneQuery = (JSONObject) JSON.toJSON(oneQueryObject);
            }

            switch (oneQuery.getString("queryMethod")) {
                case "get":
                    avQuery.getInBackground(oneQuery.getString("arg1")).subscribe(new Observer<AVObject>() {
                        @Override
                        public void onSubscribe(Disposable disposable) {}

                        @Override
                        public void onNext(AVObject avObject) {
                            result.success(avObject.toJSONObject().toJSONString());
                        }

                        @Override
                        public void onError(Throwable throwable) {
                            result.error("leancloud-error", throwable.getMessage(), null);
                        }

                        @Override
                        public void onComplete() {}
                    });
                    return;
                case "equalTo":
                    avQuery.whereEqualTo(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "notEqualTo":
                    avQuery.whereNotEqualTo(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "greaterThan":
                    avQuery.whereGreaterThan(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "greaterThanOrEqualTo":
                    avQuery.whereGreaterThanOrEqualTo(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "lessThan":
                    avQuery.whereLessThan(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "lessThanOrEqualTo":
                    avQuery.whereLessThanOrEqualTo(oneQuery.getString("arg1"), oneQuery.get("arg2"));
                    break;
                case "orderByAscending":
                    avQuery.orderByAscending(oneQuery.getString("arg1"));
                    break;
                case "orderByDescending":
                    avQuery.orderByDescending(oneQuery.getString("arg1"));
                    break;
                case "addAscendingOrder":
                    avQuery.addAscendingOrder(oneQuery.getString("arg1"));
                    break;
                case "addDescendingOrder":
                    avQuery.addDescendingOrder(oneQuery.getString("arg1"));
                    break;
                case "limit":
                    avQuery.limit(oneQuery.getIntValue("arg1"));
                    break;
                case "skip":
                    avQuery.skip(oneQuery.getIntValue("arg1"));
                    break;
                default:
                    result.error("params-error", "no such query queryMethod name, please check again", null);
                    break;
            }

        }
        avQuery.findInBackground().subscribe(new Observer<List<AVObject>>() {
            @Override
            public void onSubscribe(Disposable disposable) {}

            @Override
            public void onNext(List<AVObject> avObjects) {
                JSONObject jsonObject = new JSONObject();
                JSONArray jsonArray = new JSONArray();
                for (AVObject object : avObjects) {
                    jsonArray.add(object.toJSONObject().toJSONString());
                }
                jsonObject.put("objects", jsonArray);
                result.success(jsonObject.toJSONString());
            }

            @Override
            public void onError(Throwable throwable) {
                result.error("leancloud-error", throwable.getMessage(), null);
            }

            @Override
            public void onComplete() {}
        });
    }

    void doCloudQuery(MethodCall call, final MethodChannel.Result result) {
        String cqlString = LeancloudArgsConverter.getStringValue(call, result, "cql");
        AVQuery.doCloudQueryInBackground(cqlString).subscribe(new Observer<AVCloudQueryResult>() {
            @Override
            public void onSubscribe(Disposable disposable) {}

            @Override
            public void onNext(AVCloudQueryResult avCloudQueryResult) {
                List<? extends AVObject> avObjects = avCloudQueryResult.getResults();
                JSONObject jsonObject = new JSONObject();
                JSONArray jsonArray = new JSONArray();
                for (AVObject object : avObjects) {
                    jsonArray.add(object.toJSONObject().toJSONString());
                }
                jsonObject.put("objects", jsonArray);
                result.success(jsonObject.toJSONString());
            }

            @Override
            public void onError(Throwable throwable) {
                result.error("leancloud-error", throwable.getMessage(), null);
            }

            @Override
            public void onComplete() {}
        });
    }
}
